### PR TITLE
CellDescriptors + Tests

### PR DESCRIPTION
**Pull request by aleksei-bobrov**

### CODE REVIEW

File: Cartfile.private [hunk 1]
```diff
@@ -0,0 +1,2 @@
+github "Quick/Quick"
+github "Quick/Nimble"
```
Comment (stephanecopin) [inline Cartfile.private]: 👍

File: DataSource/DataSourceCellDescriptor.swift [hunk 2]
```diff
@@ -1,11 +1,11 @@
 import UIKit
 
-struct CellDescriptor {
-	let reuseIdentifier: String
-	let prototypeSource: PrototypeSource
-	let isMatching: (IndexPath, Any) -> Bool
+public struct CellDescriptor {
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: 🙄

File: DataSourceTests/QuickSpecWithDataSets.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//
+
+import UIKit
+import XCTest
+import DataSource
+import ReactiveSwift
+import Quick
+import Nimble
+
+class QuickSpecWithDataSets: QuickSpec {
```
Comment (stephanecopin) [inline DataSourceTests/QuickSpecWithDataSets.swift]: 👍

File: DataSourceTests/TableViewDataSourceWithHeaderFooterViewsTests.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+			let dataSource = Property(value: StaticDataSource(sections: [headerSection]))
+			tableViewDataSource = TableViewDataSourceWithHeaderFooterViews()
+			tableView = UITableView(frame: CGRect.zero)
+			let tableViewDescriptors = [CellDescriptor(TestTableViewCell.reuseIdentifier, TestCellModel.self, .class(TestTableViewCell.self))]
+			tableViewDataSource.configure(tableView, using: tableViewDescriptors)
+			tableViewDataSource.dataSource.innerDataSource <~ dataSource.producer.map { $0 as DataSource }
+		}
+		itBehavesLike("TableViewDataSource object") { ["tableViewDataSource": tableViewDataSource, "TestCellModels": self.dataSetWithTestCellModels, "tableView": tableView] }
+		it("has header") {
+			_ = expect(tableViewDataSource.tableView(tableView, viewForHeaderInSection: 0))
```
Comment (stephanecopin) [inline DataSourceTests/TableViewDataSourceWithHeaderFooterViewsTests.swift]: What should this do? Did you mean to use `expect().toNot(beNil())`?

File: DataSourceTests/TableViewDataSourceSharedConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+				}
+				it("has 1 section") {
+					expect(tableViewDataSource.numberOfSections(in: tableView)) == 1
+				}
+				it("has proper count of items") {
+					expect(tableViewDataSource.tableView(tableView, numberOfRowsInSection: 0)) == initialData.count
+				}
+				it("cells has correct type") {
+					for (index, _) in initialData.enumerated() {
+						_ = expect(tableViewDataSource.tableView(tableView, cellForRowAt: IndexPath(item: index, section: 0)))
```
Comment (stephanecopin) [inline DataSourceTests/TableViewDataSourceSharedConfiguration.swift]: Same here

File: DataSourceTests/TableViewDataSourceWithHeaderFooterViewsTests.swift (hunk 4 above)
Comment (aleksei-bobrov) [reply DataSourceTests/TableViewDataSourceWithHeaderFooterViewsTests.swift]: Yes, I meant this 👍

File: DataSourceTests/TableViewDataSourceSharedConfiguration.swift (hunk 5 above)
Comment (aleksei-bobrov) [reply DataSourceTests/TableViewDataSourceSharedConfiguration.swift]: Yes, I meant this 👍

File: DataSource/DataSourceCellDescriptor.swift [hunk 6]
```diff
@@ -56,7 +78,7 @@ extension CollectionViewDataSource {
 }
 
 extension TableViewDataSource {
-	func configure(_ tableView: UITableView, using cellDescriptors: [CellDescriptor]) {
+	public func configure(_ tableView: UITableView, using cellDescriptors: [CellDescriptor]) {
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: Same

File: DataSource/DataSourceCellDescriptor.swift [hunk 7]
```diff
@@ -77,3 +99,60 @@ extension TableViewDataSource {
 		self.tableView = tableView
 	}
 }
+
+
+extension TableViewDataSourceWithHeaderFooterViews {
+	public func configure(_ tableView: UITableView, using cellDescriptors: [CellDescriptor], headerDescriptor: HeaderFooterDescriptor?, footerDescriptor: HeaderFooterDescriptor?) {
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: Same

File: DataSource/DataSourceCellDescriptor.swift [hunk 8]
```diff
@@ -77,3 +99,60 @@ extension TableViewDataSource {
...
+
+		tableView.dataSource = self
+		self.tableView = tableView
+	}
+}
+
+extension UIView {
+	
+	public static var reuseIdentifier: String { return String(describing: self).components(separatedBy: ".").last! }
+	public static var nib: UINib { return UINib(nibName: self.reuseIdentifier, bundle: nil) }
```
[the remark this answers is not in the document]
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: Yes, this looks much better 👍 
We have to use `static var` and in the particular implementation of the protocol can be overridden using a `class var`.
And we can't put `@objc` modifier to the extensions as this modifier only works for methods of a class (not for protocols)

File: DataSource/DataSourceCellDescriptor.swift [hunk 9]
```diff
@@ -24,16 +24,38 @@ struct CellDescriptor {
...
 
+public enum PrototypeSource {
+	case storyboard
+	case nib(UINib)
+	case `class`(AnyObject.Type)
+}
+
 extension CollectionViewDataSource {
-	func configure(_ collectionView: UICollectionView, using cellDescriptors: [CellDescriptor]) {
+	public func configure(_ collectionView: UICollectionView, using cellDescriptors: [CellDescriptor]) {
```
[the remark this answers is not in the document]
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: 💯

File: DataSource/DataSourceCellDescriptor.swift (hunk 6 above)
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: 💯

File: DataSource/DataSourceCellDescriptor.swift (hunk 7 above)
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: 💯

File: DataSourceTests/QuickSpecWithDataSets.swift [hunk 10]
```diff
@@ -44,4 +45,11 @@ class QuickSpecWithDataSets: QuickSpec {
 
 	let dataSetWithTestCellModels = [TestCellModel(), TestCellModel(), TestCellModel()]
 	let dataSetWithTestCellModels2 = [TestCellModel()]
+
+	var coreDataManager: CoreDataManager?
```
Comment (stephanecopin) [inline DataSourceTests/QuickSpecWithDataSets.swift]: This probably should be an implicitely unwrapped optional:
```suggestion
	var coreDataManager: CoreDataManager!
```

**Comment by stephanecopin** [comment]
Looks good, LGTM

**Pull request merged**